Automattic/pocket-casts-android
# Add share action to swipe menus

## Description

This adds a share option to the swipe menus as requested in the [Increase "Share" discoverability](https://github.com/Automattic/pocket-casts-android/discussions/1168) discussion. 

Because our swipe menus are custom made to handle up to two icons on the left and only a single icon on the right, adding a second icon to the right was a bit tricky. 

In addition, the way this was set up before was kind of complex because the click listener for the buttons came from a completely different place from where the icon and background color was set. This made it very difficult to avoid bugs due to these two sources of information getting out of sync. After spending a day or so playing whack-a-mole fixing a bug, and then fixing a new bug the previous fix caused, and then fixing a new bug the previous..., etc., I decided to do a broader refactor. Now a lot more of the swipe button logic is consolidated in the `SwipeButtonLayoutViewModel`.

I think this consolidates the logic around the share buttons nicely, but it does not consolidate the view handling itself—everywhere we want to use a swipe layout still requires a fair bit of manual XML work, but addressing that didn't seem worth it because (1) I already spent more time on this than I wanted and (2) the whole app will be written in compose soon anyway. 😉 

![Screenshot 2023-07-21 at 2 29 46 PM](https://github.com/Automattic/pocket-casts-android/assets/4656348/17fca221-8294-49dc-9733-573b2454d6ef)

Related iOS PR: https://github.com/Automattic/pocket-casts-ios/pull/966

## Testing Instructions

Because of the refactoring, when testing this PR, we should test all the possible swipe actions, not just the share action. Apart from adding the share button to the Podcast, Filter, Downloads, Starred, and Listening History screens, all other swipe behavior should be unchanged.

Some things to watch out for include:
1. Make sure that when you release the swipe menu so the buttons remain visible, that the buttons have the right width (had a lot of bugs where it wasn't 1-button-wide when showing 1 button and 2-buttons-wide when showing two buttons).
2. Do multiple actions that change the buttons on a single row (i.e., adding and removing the same episode from the up next queue) to make sure the buttons up date appropriately for the episode's new state.

### 1. Podcast, Filter, Downloads, Starred, and Listening History:

* **For unqueued episodes:** LtR swipes reveal the add to up next options. The left button should be the default that user has selected under the General settings. Doing a full-swipe should apply the left button. The only way to apply the right button is the menu open a small amount, release the swipe so the buttons remain visible, and tap the right button.
* **For queued episodes:** LtR swipes reveal the remove from queue option.
* **For archived episodes:** RtL swipes reveal the share and archive buttons. The archive button is on the right and is the action that is taken with a full swipe.
* **For unarchived episodes:** RtL swipes are the same as for archived episodes except the archive action is changed to an unarchived action.

### 2. Files

The Files screen's swipe behavior should be unchanged by this PR. 
* LtR swipes will reveal the add to up next options (ordered based on the user's setting). 
* RtL swipes will reveal only the delete option.

### 3. Up Next Queue

* LtR swipes will always reveal the add-to-queue (or in this case move-in-queue) buttons. Unlike on the other screens, the order here never changes. The left button is move-to-top and the right button is move-to-bottom regardless of what the user's setting says should be the default action.
* RtL swipes reveal the remove from queue button

### 4. Analytics
* For each swipe action, we should send an appropriate`episode_swipe_action_performed` with `action` and `source`, and `episode_uuid` properties. 
* When sharing an episode from the swipe menu, the `podcast_shared` event is sent with a `source` of "swipe_action"
* I have added the `episode_uuid` as a property to the `episode_*_to_up_next` events since we were missing that, but it was identified on the tracking document

## Checklist
 
#### I have tested any UI changes...
